runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/ProcessManager.java

```diff
@@ -0,0 +1,157 @@
+import com.google.common.collect.ImmutableList;
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.concurrent.ThreadSafe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** A simple process manager which forks processes and kills them if necessary. */
+@ThreadSafe
+class ProcessManager {
+  private static final Logger LOG = LoggerFactory.getLogger(ProcessManager.class);
+
+  private static final ProcessManager INSTANCE = new ProcessManager();
+
+  private final Map<String, Process> processes;
+
+  public static ProcessManager getInstance() {
+    return INSTANCE;
+  }
+
+  private ProcessManager() {
+    this.processes = Collections.synchronizedMap(new HashMap<>());
+  }
+
+  static class RunningProcess {
+    private Process process;
+
+    RunningProcess(Process process) {
+      this.process = process;
+    }
+
+    /** Checks if the underlying process is still running. */
+    void isAliveOrThrow() throws IllegalStateException {
+      if (!process.isAlive()) {
+        throw new IllegalStateException("Process died with exit code " + process.exitValue());
+      }
+    }
+
+    @VisibleForTesting
+    Process getUnderlyingProcess() {
+      return process;
+    }
+  }
+
+  /**
+   * Forks a process with the given command and arguments.
+   *
+   * @param id A unique id for the process
+   * @param command the name of the executable to run
+   * @param args arguments to provide to the executable
+   * @return A RunningProcess which can be checked for liveness
+   */
+  RunningProcess startProcess(String id, String command, List<String> args) throws IOException {
+    return startProcess(id, command, args, Collections.emptyMap());
+  }
+
+  /**
+   * Forks a process with the given command, arguments, and additional environment variables.
+   *
+   * @param id A unique id for the process
+   * @param command The name of the executable to run
+   * @param args Arguments to provide to the executable
+   * @param env Additional environment variables for the process to be forked
+   * @return A RunningProcess which can be checked for liveness
+   */
+  RunningProcess startProcess(String id, String command, List<String> args, Map<String, String> env)
+      throws IOException {
+    checkNotNull(id, "Process id must not be null");
+    checkNotNull(command, "Command must not be null");
+    checkNotNull(args, "Process args must not be null");
+    checkNotNull(env, "Environment map must not be null");
+
+    ProcessBuilder pb =
+        new ProcessBuilder(ImmutableList.<String>builder().add(command).addAll(args).build());
+    pb.environment().putAll(env);
+
+    LOG.debug("Attempting to start process with command: " + pb.command());
+    // Pipe stdout and stderr to /dev/null to avoid blocking the process due to filled PIPE buffer
+    pb.redirectErrorStream(true);
```

Comment:
Consider inheriting stderr (and possibly also stdout), possibly optional, since it may be hard to diagnose issues when the output isn't available.